controller/internal/enforcer/envoyauthorizer/envoyproxy/auth_server.go

```diff
@@ -0,0 +1,466 @@
+package envoyproxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	ext_auth "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	envoy_type "github.com/envoyproxy/go-control-plane/envoy/type"
+	"go.aporeto.io/trireme-lib/collector"
+	"go.aporeto.io/trireme-lib/common"
+	"go.aporeto.io/trireme-lib/controller/internal/enforcer/apiauth"
+	"go.aporeto.io/trireme-lib/controller/internal/enforcer/applicationproxy/serviceregistry"
+	"go.aporeto.io/trireme-lib/controller/internal/enforcer/flowstats"
+	"go.aporeto.io/trireme-lib/controller/internal/enforcer/metadata"
+	"go.aporeto.io/trireme-lib/controller/pkg/secrets"
+	"go.aporeto.io/trireme-lib/controller/pkg/servicetokens"
+	"go.aporeto.io/trireme-lib/utils/cache"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	rpc "istio.io/gogo-genproto/googleapis/google/rpc"
+)
+
+const (
+	// IngressSocketPath is the unix socket path where the authz server will be listening on for the ingress authz server
+	//IngressSocketPath = "@aporeto_envoy_authz_ingress"
+	IngressSocketPath = "127.0.0.1:1999"
+
+	// EgressSocketPath is the unix socket path where the authz server will be listening on for the egress authz server
+	EgressSocketPath = "127.0.0.1:1998"
+	//EgressSocketPath = "@aporeto_envoy_authz_egress"
+
+	// defaultValidity of the issued JWT token
+	defaultValidity = 60 * time.Second
+
+	// aporetoKeyHeader is the HTTP header name for the key header
+	aporetoKeyHeader = "x-aporeto-key"
+
+	// aporetoAuthHeader is the HTTP header name for the auth header
+	aporetoAuthHeader = "x-aporeto-auth"
+)
+
+// Direction is used to indicate if the authorization server is ingress or egress.
+// NOTE: the type is currently set to uint8 and not bool because in Istio there are 3 types:
+// - SIDECAR_INBOUND
+// - SIDECAR_OUTBOUND
+// - GATEWAY
+// And we are not sure yet if we need an extra authz server for GATEWAY.
+type Direction uint8
+
+const (
+	// UnknownDirection is only used to denote uninitialized variables
+	UnknownDirection Direction = 0
+
+	// IngressDirection refers to inbound / ingress traffic.
+	// NOTE: for Istio use this in conjunction with SIDECAR_INBOUND
+	IngressDirection Direction = 1
+
+	// EgressDirection refers to outbound / egress traffic.
+	// NOTE: for Istio use this in conjunction with SIDECAR_OUTBOUND
+	EgressDirection Direction = 2
+)
+
+// String overwrites the string interface
+func (d Direction) String() string {
+	switch d {
+	case UnknownDirection:
+		return "UnknownDirection"
+	case IngressDirection:
+		return "IngressDirection"
+	case EgressDirection:
+		return "EgressDirection"
+	default:
+		return fmt.Sprintf("Unimplemented(%d)", d)
+	}
+}
+
+// AuthServer struct, the server to hold the envoy External Auth.
+type AuthServer struct {
+	puID       string
+	puContexts cache.DataStore
+	secrets    secrets.Secrets
```

Comment:
should probably be `LockedSecrets` so that you don't have to update the secrets in every auth server ... at least I think that is what I planned there when I remember correctly :)

Reply:
yeah but, the secrets is used by api auth now, need to talk to dimitri about it. As the apiauth is using the normal secrets. I will talk to him to see if we can make it to Lockedsecrets.